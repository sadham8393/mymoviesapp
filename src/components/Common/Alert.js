import React from 'react';
import PropTypes from 'prop-types';

const Alert = (props) =>{
    return (
        <nav className = {`${props.alertClass} alert-container`}>
            <div className="nav-wrapper">
            <ul className="left">
                <h5>{props.alertMessage}</h5>
            </ul>
            <ul className="right">
                <i className="small material-icons alert-close-btn " onClick = {props.alertClose}> close </i>
            </ul>
            </div>
        </nav>
    )
}

Alert.propTypes = { 
    alertClass: PropTypes.string, 
    alertMessage: PropTypes.string,
    alertClose: PropTypes.func
} 

export default Alert;