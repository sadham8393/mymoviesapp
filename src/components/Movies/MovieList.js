import React from 'react';
import Movie from './Movie';
import PropTypes from 'prop-types';

const MovieList = (props) =>{
    return (
        <div className ="container movies-container">
            <div className = "row">
                <div className = "col s12">
                    {
                        props.movies.map((movie, i ) => {
                            return (
                                <Movie key={i} movie = {movie} movieTitle = {movie.original_title} 
                                            image = {movie.poster_path} viewInfoModal = {props.viewInfoModal} onDeleteClick = {props.onDeleteClick}/>
                            )
                        })
                    }
                </div>
            </div>
        </div>
    )
}

MovieList.propTypes = { 
    title: PropTypes.string, 
    onDeleteClick: PropTypes.func,
    viewInfoModal : PropTypes.func
} 

export default MovieList;