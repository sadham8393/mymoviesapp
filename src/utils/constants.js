// URl constants

export const MOVIEDB_PATH = "https://api.themoviedb.org/3/";
export const REACT_MOVIEDB_API_KEY  = "?api_key=0e89680c8b3fa0f259b53bbcc8f6091d";
export const MOVIEDB_IMAGE_APPEND_URL = "https://image.tmdb.org/t/p/w500/";
export const TRENDING_DAY_URL = "trending/all/day";


export const NO_IMAGE_URl = "https://s3-ap-southeast-1.amazonaws.com/upcode/static/default-image.jpg";

//Common Constants
export const INPUT_TYPE_TEXT = "text";
export const INPUT_TYPE_SELECT = "select";
export const NOT_AVAILABLE = "Not Available";


//Action Constants
export const FETCH_TRENDING_MOVIES = "FETCH_TRENDING_MOVIES";
export const FETCH_TV_MOVIES = "FETCH_TV_MOVIES";
export const DELETE_TV_MOVIES = "DELETE_TV_MOVIES";
export const MOVIES_FETCHED_SUCCESS = "MOVIES_FETCHED_SUCCESS";
export const MOVIES_FETCHED_ERROR = "MOVIES_FETCHED_ERROR";
export const DELETE_MOVIE_SUCCESS = "DELETE_MOVIE_SUCCESS";
export const DELETE_MOVIE_FAILURE = "DELETE_MOVIE_FAILURE";

export const MOVIES_FETCHED_NO_RESULTS = "MOVIES_FETCHED_NO_RESULTS";

export const RESET_RESPONSE = "RESET_RESPONSE";

export const NO_RESULTS_FOUND = "No results found for the given search input";

export const SERVER_UNAVAILABE = "Server unavailable. Please try after sometime"



