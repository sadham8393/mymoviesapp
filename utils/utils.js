import configureStore from 'redux-mock-store'

export const findByTestClass = (component, className) => {
    const wrapper = component.find(`${className}`);
    return wrapper;
};

export const findByComponent = (component, innerComponent) => {
    const wrapper = component.find(innerComponent);
    return wrapper;
};

export const testStore = (initialState) => {
    let store;
    const mockStore = configureStore();
    store = mockStore(initialState)
    return store;
};