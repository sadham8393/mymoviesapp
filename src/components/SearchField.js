import React from 'react';
import PropTypes from 'prop-types';

import {INPUT_TYPE_TEXT, INPUT_TYPE_SELECT} from '../utils/constants'

const SearchField = (props) =>{
    return (
        <div className ="container search-container">
            <form action = "" onSubmit = {props.handleSearch} className="row" >
                <div className="input-field col s4">
                    <input placeholder="Search Movie/Tv" className = "validate" id="searchInput" type="text" 
                            name = "searchterm" onChange = {(e) => props.handleChange(e, INPUT_TYPE_TEXT)} required/>
                    <span className="helper-text" data-error="Please enter Movie/Tv Series and Search." ></span>
                </div>
                
                <div className="input-field col s2">
                    <select className="browser-default" name = "tvMovieType" onChange = {(e) => props.handleChange(e, INPUT_TYPE_SELECT)}>
                    <option value="" disabled defaultValue>Search By</option>
                    <option value="multi">All</option>
                    <option value="movie">Movie</option>
                    <option value="tv">TV</option>
                    </select>
                </div>

                <div className="input-field col s2">
                    <button className="waves-effect waves-light btn" disabled = {props.searchBtnDisabled} onClick = {props.handleSearch}><i className="material-icons right">search</i>Search</button>
                </div>
            </form>
        </div>
    )
}

SearchField.propTypes = { 
    handleSearch: PropTypes.func,
    handleChange : PropTypes.func,
    searchBtnDisabled : PropTypes.bool

} 
export default SearchField;