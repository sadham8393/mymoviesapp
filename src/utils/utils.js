import { NOT_AVAILABLE } from './constants';

/**
 * Convert date format from "yyyy-mm-dd" to "dd/mm/yy"
 */

export function convertDate(date){
    if(date){
        return date.toString().split("-").reverse().join("/");
    }
    return NOT_AVAILABLE;
}