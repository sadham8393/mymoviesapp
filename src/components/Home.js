import React, {Component} from 'react';
import { INPUT_TYPE_TEXT } from '../utils/constants';
import { deleteMovieMessage } from '../utils/utils';
import { connect } from 'react-redux';
import SearchField from './SearchField';
import Header from './Nav';
import Pagination from './Common/Pagination';
import MovieList from './Movies/MovieList';
import MovieInfoModal from './Movies/MovieInfoModal';
import * as actions from '../actions/moviesAction';
import Alert from './Common/Alert';
import Loader from './Common/Loader';
import ConfirmModal from './Common/ConfirmModal';

class Home extends Component {
    constructor(){
        super();
        this.state = {
          initialLoading : true,
          moviesList :{},
          movies : [],
          searchterm : "",
          tvMovieType : "movie",
          searchBtnDisabled : true,
          currentPage : 1,
          pageCount : 0,
          totalCount :0,
          fieldValid : true,
          message : "",
          alertClass : "hide",
          isLoading : false,
          infoModalOpen : false,
          currentMovie : null,
          confirmModalOpen : false,
          movieId : "", 
          movieName : ""
        }
    }

    componentDidMount(){
      this.props.getTrendingList(1);
    }
    componentDidUpdate(){
      window.scrollTo(0, 0);
    }

    static getDerivedStateFromProps(nextProps, prevState){
      let moviesList = nextProps.moviesList,
          infoMessage = nextProps.info,
          errorMessage = nextProps.error,
          successMessage = nextProps.success,
          isLoading = nextProps.isLoading;
      if(isLoading){
        return {
          isLoading : isLoading
        };
      }
      if(infoMessage){
        return {
          moviesList : {},
          movies : [], 
          pageCount : 0,
          currentPage : 1, 
          totalCount : 0,
          message : infoMessage, alertClass : "show info", isLoading : false
        }
      } 
      if(errorMessage) {
        return {
          moviesList : {},
          movies : [], 
          pageCount : 0,
          currentPage : 1, 
          totalCount : 0,
          message : errorMessage, alertClass : "show error", isLoading : false
        }
      }

      if(moviesList && moviesList.results.length >= 0) {
        return {
          moviesList : moviesList,
          movies : [...moviesList.results], 
          pageCount : moviesList.total_pages,
          currentPage : moviesList.page, 
          totalCount : moviesList.total_results,
          isLoading : false,
          message : successMessage ? successMessage : "" , 
          alertClass : successMessage ? "show success" : "hide"
        };
      }
      
      return prevState;
    }

    handleSearch = (e) => {
      this.closeAlert();
      e.preventDefault();
      let { tvMovieType, searchterm } = this.state;
      if(!searchterm){
        this.setState({ fieldValid : false});
      }
      this.setState({ initialLoading:false});
      this.props.getMoviesList({tvMovieType, searchterm, pageNumber : 1 });
    }

    nextPage = (pageNumber) => {
      this.closeAlert();
      let {tvMovieType, searchterm } = this.state;
      if(this.state.initialLoading){
        this.props.getTrendingList(pageNumber);
      } else {
        this.props.getMoviesList({tvMovieType, searchterm, pageNumber });
      } 
    }

    handleChange = (e, fieldType) => {
      this.closeAlert();
      let input = e.target.value;
      if(INPUT_TYPE_TEXT === fieldType && input ){
        this.setState({ searchBtnDisabled : false});
      } else if(INPUT_TYPE_TEXT === fieldType && !input) {
        this.setState({ searchBtnDisabled : true});
      }
      this.setState({[e.target.name] : input ? input : "" , currentPage : 1});
    }

    headerClick = () => {
      this.closeAlert();
      this.setState(prevState => ({
        initialLoading: !prevState.initialLoading
      }));
      this.props.resetResponse();
      this.props.getTrendingList(1);
    }

    renderPagination = (totalPages) => {
      if(totalPages > 1){
        return (
          <Pagination totalCount = {this.state.totalCount} pages = {totalPages} nextPage = {this.nextPage} currentPage = {this.state.currentPage}/>
        )
      }
    }
    
    closeAlert = () => {
      this.props.resetResponse();
      this.setState({alertClass : "hide"})
    }

    renderLoader = () => {
      if(this.state.isLoading){
        return (
          <Loader />
        )
      }
    }

    renderInfoModal = () => {
      if(this.state.infoModalOpen){
          return (
            <MovieInfoModal movie = {this.state.currentMovie} viewInfoModalOpen = {this.state.infoModalOpen} closeInfoModal={this.closeInfoModal}/>
          )
      }
    }
    renderConfirmModal = () =>{
      if(this.state.confirmModalOpen){
          return (
            <ConfirmModal confirmModalOpen = {this.state.confirmModalOpen} confirmMessage = {deleteMovieMessage(this.state.movieName)} 
            title = {`Delete ${this.state.movieName}`} onYesClick = {this.confirmYesClick} onNoClick = {this.confirmNoClick}/>
          )
      }
    }

    confirmYesClick = () => {
      let moviesList = this.state.moviesList,
          movieId = this.state.movieId;
          this.props.deleteTvMovie(movieId, moviesList);
      this.setState(prevState => ({
        confirmModalOpen: !prevState.confirmModalOpen, movieId : "", movieName : ""
      }));
    }

    confirmNoClick = () => {
      this.setState(prevState => ({
        confirmModalOpen: !prevState.confirmModalOpen, movieId : "", movieName : ""
      }));
    }

    onDeleteClick = (movie) => {
      let movieId = movie.id,
          movieName = movie.name || movie.title;
      this.setState(prevState => ({
        confirmModalOpen: !prevState.confirmModalOpen, movieId : movieId, movieName : movieName
      }));
    }

    viewInfoModal = (currentMovieDetails) => {
        this.setState({
          infoModalOpen: true, currentMovie : currentMovieDetails
        });
    }

    closeInfoModal = () => {
      this.setState({
        infoModalOpen: false, currentMovie : null
      });
    }

    render() {
      let totalPages = this.state.pageCount,
          alertMessage = this.state.message;
      return (
        <div className = "home-container">
            {this.renderLoader()}
            {this.renderInfoModal()}
            {this.renderConfirmModal()}
            <Header onHeaderClick = {this.headerClick}/>
            <Alert alertClass = {this.state.alertClass} alertClose = {this.closeAlert} alertMessage = {alertMessage}/>
            <SearchField handleSearch = {this.handleSearch} handleChange = {this.handleChange} 
                          searchBtnDisabled = {this.state.searchBtnDisabled} fieldValid = {this.state.fieldValid}/>
            {this.renderPagination(totalPages)}
            <MovieList movies = {this.state.movies} viewInfoModal = {this.viewInfoModal} onDeleteClick = {this.onDeleteClick}/>
            {this.renderPagination(totalPages)}
            
        </div>
      );
    }
}

const mapStateToProps = (state) => ({
  moviesList : state.movies.moviesList,
  info : state.movies.info,
  error : state.movies.error,
  success : state.movies.success,
  isLoading : state.movies.isLoading
})

export default connect(mapStateToProps,actions)(Home);