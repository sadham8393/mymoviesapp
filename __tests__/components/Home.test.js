import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { findByTestClass, findByComponent , testStore } from '../../utils/utils';
import Home from '../../src/components/Home';
import SearchField from '../../src/components/SearchField';
import Header from '../../src/components/Nav';
import MovieList from '../../src/components/Movies/MovieList';
import Alert from '../../src/components/Common/Alert';
import { INPUT_TYPE_TEXT, INPUT_TYPE_SELECT } from '../../src/utils/constants';

const setUp = (initialState={}) => {
    const store = testStore(initialState);
    const wrapper = shallow(<Home store={store} />).childAt(0).dive();;
    return wrapper;
};

describe('<Home /> component', () => { 
    let wrapper,
        initialState;
    const event = {
        preventDefault() {},
        target :{value : "Test"}
    };
    let fieldType = INPUT_TYPE_TEXT;
    beforeEach(() => {
        initialState = {movies :{
            isLoading : false,
            moviesList: {
                "page": 1,
                "results": [
                {
                    "adult": false,
                    "backdrop_path": "/bOGkgRGdhrBYJSLpXaxhXVstddV.jpg",
                    "genre_ids": [ 28, 12, 14, 878],
                    "id": 299536,
                    "original_language": "en",
                    "original_title": "Avengers: Infinity War",
                    "overview": "As the Avengers and their allies have continued to protect the world from threats too large for any one hero to handle, a new danger has emerged from the cosmic shadows: Thanos. A despot of intergalactic infamy, his goal is to collect all six Infinity Stones, artifacts of unimaginable power, and use them to inflict his twisted will on all of reality. Everything the Avengers have fought for has led up to this moment - the fate of Earth and existence itself has never been more uncertain.",
                    "poster_path": "/7WsyChQLEftFiDOVTGkv3hFpyyt.jpg",
                    "release_date": "2018-04-25",
                    "title": "Avengers: Infinity War",
                    "video": false,
                    "vote_average": 8.3,
                    "vote_count": 6937,
                    "popularity": 358.799
                }
                ],
                "total_pages": 1,
                "total_results": 1
            }
          }};
          wrapper = setUp(initialState);
    });

    it('should render correctly', () => {
        const store = testStore(initialState);
        const component = renderer.create(<Home store={store} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Should render without errors', () => {
        const component = findByTestClass(wrapper, '.home-container');
        expect(component.length).toBe(1);
    });

    it('should render search field and length to be 1', () => {
        const component = findByComponent(wrapper, SearchField);
        expect(component.length).toBe(1);
    });

    it('should render Header and length to be 1', () => {
        const component = findByComponent(wrapper, Header);
        expect(component.length).toBe(1);
    });
    
    it('should render MovieList and length to be 1', () => {
        const component = findByComponent(wrapper, MovieList);
        expect(component.length).toBe(1);
    });

    it('should render Alert and length to be 1', () => {
        const component = findByComponent(wrapper, Alert);
        expect(component.length).toBe(1);
    });

    it('handleSearch Method should update state as expected', () => {
        const classInstance = wrapper.instance();
        classInstance.handleSearch(event);
        const newState = classInstance.state.fieldValid;
        expect(newState).toBe(false);
    });

    it('handleSearch Method should update state as expected', () => {
        const classInstance = wrapper.instance();
        classInstance.handleSearch(event);
        const newState = classInstance.state.initialLoading;
        expect(newState).toBe(false);
    });

    it('handleChange Method should update state as expected if the field type is text', () => {
        const classInstance = wrapper.instance();
        classInstance.handleChange(event, fieldType);
        const newState = classInstance.state.searchBtnDisabled;
        expect(newState).toBe(false);
    });

    it('handleChange Method should update state as expected if the field type is select', () => {
        fieldType = INPUT_TYPE_SELECT;
        const classInstance = wrapper.instance();
        classInstance.handleChange(event, fieldType);
        const newState = classInstance.state.searchBtnDisabled;
        expect(newState).toBe(true);
    });

    it('closeAlert Method should update alertClass in state', () => {
        const classInstance = wrapper.instance();
        classInstance.closeAlert();
        const newState = classInstance.state.alertClass;
        expect(newState).toBe("hide");
    });

    it('headerClick Method should update initialLoading in state', () => {
        const classInstance = wrapper.instance();
        classInstance.headerClick();
        classInstance.state.initialLoading = false;
        const newState = !classInstance.state.initialLoading;
        expect(newState).toBe(true);
    });

    it('viewInfoModal Method should update infoModalOpen in state', () => {
        const classInstance = wrapper.instance(),
              currentMovieDetails = {};
        classInstance.viewInfoModal(currentMovieDetails);
        const newState = classInstance.state.infoModalOpen;
        expect(newState).toBe(true);
    });

    it('viewInfoModal Method should update currentMovie in state', () => {
        const classInstance = wrapper.instance(),
              currentMovieDetails = {movieName : "Avengers"};
        classInstance.viewInfoModal(currentMovieDetails);
        const newState = classInstance.state.currentMovie;
        expect(newState).toBe(currentMovieDetails);
    });

    it('closeInfoModal Method should update infoModalOpen in state', () => {
        const classInstance = wrapper.instance();
        classInstance.closeInfoModal();
        const newState = classInstance.state.infoModalOpen;
        expect(newState).toBe(false);
    });
    it('closeInfoModal Method should update currentMovie in state', () => {
        const classInstance = wrapper.instance();
        classInstance.closeInfoModal();
        const newState = classInstance.state.currentMovie;
        expect(newState).toBe(null);
    });
    
});