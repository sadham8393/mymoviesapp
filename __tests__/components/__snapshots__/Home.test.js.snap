// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> component should render correctly 1`] = `
<div
  className="home-container"
>
  <nav
    className="navbar"
  >
    <div
      className="nav-wrapper nav-header"
    >
      <h2
        className="brand-logo"
        onClick={[Function]}
      >
         My Movies App 
      </h2>
    </div>
  </nav>
  <nav
    className="hide alert-container"
  >
    <div
      className="nav-wrapper"
    >
      <ul
        className="left"
      >
        <h5>
          
        </h5>
      </ul>
      <ul
        className="right"
      >
        <i
          className="small material-icons alert-close-btn "
          onClick={[Function]}
        >
           close 
        </i>
      </ul>
    </div>
  </nav>
  <div
    className="container search-container"
  >
    <form
      action=""
      className="row"
      onSubmit={[Function]}
    >
      <div
        className="input-field col s4"
      >
        <input
          className="validate"
          id="searchInput"
          name="searchterm"
          onChange={[Function]}
          placeholder="Search Movie/Tv"
          required={true}
          type="text"
        />
        <span
          className="helper-text"
          data-error="Please enter Movie/Tv Series and Search."
        />
      </div>
      <div
        className="input-field col s2"
      >
        <select
          className="browser-default"
          name="tvMovieType"
          onChange={[Function]}
        >
          <option
            defaultValue={true}
            disabled={true}
            value=""
          >
            Search By
          </option>
          <option
            value="multi"
          >
            All
          </option>
          <option
            value="movie"
          >
            Movie
          </option>
          <option
            value="tv"
          >
            TV
          </option>
        </select>
      </div>
      <div
        className="input-field col s2"
      >
        <button
          className="waves-effect waves-light btn"
          disabled={true}
          onClick={[Function]}
        >
          <i
            className="material-icons right"
          >
            search
          </i>
          Search
        </button>
      </div>
    </form>
  </div>
  <div
    className="container movies-container"
  >
    <div
      className="row"
    >
      <div
        className="col s12"
      >
        <div
          className="col s12 m6 l3 movies-card"
        >
          <div
            className="card"
          >
            <div
              className="card-image waves-effect waves-block waves-light"
            >
              <img
                alt="Avengers: Infinity War"
                className="movie-img"
                src="https://image.tmdb.org/t/p/w500//7WsyChQLEftFiDOVTGkv3hFpyyt.jpg"
              />
            </div>
          </div>
          <div
            className="card-content"
          >
            <button
              className="waves-effect waves-light btn-small overview"
              onClick={[Function]}
            >
                Overview 
            </button>
            <button
              className="waves-effect waves-light btn-small deleteBtn"
              onClick={[Function]}
            >
               Delete 
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
